File: sdk/python/pageros/signing.py
"""Request signature verification (PY-004).

Implements the `PagerOS-Sig` request authentication scheme defined in
SPEC.md §9.2:

    Every request to an App Server includes ``PagerOS-Sig``: an Ed25519
    signature over ``method || url || timestamp || body_hash``. Timestamp
    prevents replay (servers reject > 5 min skew).

Header layout used on the wire:

* ``PagerOS-Device``    – base64url (no padding) of the device Ed25519
                          public key (32 bytes).
* ``PagerOS-Sig``       – base64url (no padding) of the Ed25519 signature
                          (64 bytes).
* ``PagerOS-Timestamp`` – integer Unix seconds, as ASCII (matches the
                          test-vector layout in
                          ``docs/spec/crypto-test-vectors.json``).

The signing input is the byte concatenation, without separators::

    utf8(method) + utf8(url) + utf8(timestamp) + sha256(body)

where ``sha256(body)`` is the raw 32-byte digest of the request body (empty
body hashes the empty string). This matches the
``ed25519-pageros-sig-sample`` vector in the shared test-vector file.

This module deliberately stays framework-agnostic: it exposes the
low-level :func:`verify_request` and a thin WSGI middleware
(:class:`SignatureVerifierMiddleware`). Higher-level integration with the
PY-008 ``ctx`` object plugs in by reading the verified device pubkey from
the WSGI environ key ``"pageros.device_id"``.

Cryptography uses PyNaCl, per ``docs/spec/crypto-suite.md`` (SEC-001) — the
same libsodium binding the firmware links against, so signatures round-trip
byte-for-byte across implementations.
"""

from __future__ import annotations

import base64
import hashlib
import time
from dataclasses import dataclass
from typing import Any, Callable, Iterable, Mapping

from nacl.exceptions import BadSignatureError
from nacl.signing import SigningKey, VerifyKey

HEADER_DEVICE = "PagerOS-Device"
HEADER_SIG = "PagerOS-Sig"
HEADER_TIMESTAMP = "PagerOS-Timestamp"

DEFAULT_MAX_SKEW_SECONDS = 300  # SPEC §9.2: reject > 5 min skew.

ED25519_PUBKEY_LEN = 32
ED25519_SIG_LEN = 64


# --------------------------------------------------------------------------- #
# Errors
# --------------------------------------------------------------------------- #


class SignatureError(ValueError):
    """Base class for all signature-verification failures."""


class MissingHeader(SignatureError):
    """A required PagerOS-* header was missing from the request."""


class InvalidEncoding(SignatureError):
    """A PagerOS-* header could not be parsed (bad base64, bad length, ...)."""


class TimestampSkew(SignatureError):
    """The request timestamp is outside the allowed skew window."""


class BadSignature(SignatureError):
    """The signature did not verify against the device public key."""


# --------------------------------------------------------------------------- #
# Low-level helpers
# --------------------------------------------------------------------------- #


def _b64url_decode(value: str, *, expected_len: int, field: str) -> bytes:
    """Decode a base64url (no-padding) string and assert the byte length."""
    if not isinstance(value, str):
        raise InvalidEncoding(f"{field}: not a string")
    padding = "=" * (-len(value) % 4)
    try:
        raw = base64.urlsafe_b64decode(value + padding)
    except (ValueError, base64.binascii.Error) as exc:  # type: ignore[attr-defined]
        raise InvalidEncoding(f"{field}: not valid base64url") from exc
    if len(raw) != expected_len:
        raise InvalidEncoding(
            f"{field}: expected {expected_len} bytes, got {len(raw)}"
        )
    return raw


def _b64url_encode(raw: bytes) -> str:
    return base64.urlsafe_b64encode(raw).rstrip(b"=").decode("ascii")


def compute_body_hash(body: bytes | None) -> bytes:
    """Return the SHA-256 digest used in the signing input."""
    return hashlib.sha256(body or b"").digest()


def build_signing_input(
    method: str, url: str, timestamp: str | int, body: bytes | None
) -> bytes:
    """Assemble the bytes that get signed.

    Matches the ``msg_assembly`` rule of the
    ``ed25519-pageros-sig-sample`` test vector: concatenate the UTF-8
    encodings of ``method``, ``url`` and ``timestamp`` followed by the raw
    SHA-256 of the body. No separators.
    """
    ts = str(timestamp)
    return (
        method.encode("utf-8")
        + url.encode("utf-8")
        + ts.encode("utf-8")
        + compute_body_hash(body)
    )


def _get_case_insensitive(headers: Mapping[str, str], key: str) -> str | None:
    """Look up a header with case-insensitive comparison.

    The middleware is happy to accept any of ``PagerOS-Device``,
    ``pageros-device``, ``PAGEROS-DEVICE``, etc., because WSGI servers and
    HTTP clients normalize casing inconsistently.
    """
    if key in headers:
        return headers[key]
    lower = key.lower()
    for k, v in headers.items():
        if k.lower() == lower:
            return v
    return None


# --------------------------------------------------------------------------- #
# Verification
# --------------------------------------------------------------------------- #


@dataclass(frozen=True)
class VerifiedRequest:
    """Result of a successful signature check."""

    device_id: bytes
    """Raw 32-byte Ed25519 device public key (the PagerOS device identity)."""

    timestamp: int
    """The Unix-seconds timestamp embedded in the request."""

    @property
    def device_id_b64(self) -> str:
        """Convenience accessor for the base64url-encoded device id."""
        return _b64url_encode(self.device_id)


def verify_request(
    method: str,
    url: str,
    headers: Mapping[str, str],
    body: bytes | None = b"",
    *,
    max_skew_seconds: int = DEFAULT_MAX_SKEW_SECONDS,
    now: Callable[[], float] | None = None,
) -> VerifiedRequest:
    """Verify ``PagerOS-Sig`` against ``headers`` for the given request.

    Raises a :class:`SignatureError` subclass on any failure; on success
    returns a :class:`VerifiedRequest` carrying the verified device pubkey
    and timestamp.

    ``method`` and ``url`` MUST be the same byte-for-byte strings the
    client used when computing the signature. Callers integrating into a
    web framework should pass the raw request path-with-query (e.g.
    ``environ["RAW_URI"]`` for WSGI).
    """
    device_header = _get_case_insensitive(headers, HEADER_DEVICE)
    sig_header = _get_case_insensitive(headers, HEADER_SIG)
    ts_header = _get_case_insensitive(headers, HEADER_TIMESTAMP)

    if device_header is None:
        raise MissingHeader(f"missing {HEADER_DEVICE}")
    if sig_header is None:
        raise MissingHeader(f"missing {HEADER_SIG}")
    if ts_header is None:
        raise MissingHeader(f"missing {HEADER_TIMESTAMP}")

    pubkey = _b64url_decode(
        device_header, expected_len=ED25519_PUBKEY_LEN, field=HEADER_DEVICE
    )
    sig = _b64url_decode(
        sig_header, expected_len=ED25519_SIG_LEN, field=HEADER_SIG
    )

    try:
        ts_int = int(ts_header)
    except ValueError as exc:
        raise InvalidEncoding(
            f"{HEADER_TIMESTAMP}: expected integer Unix seconds, got {ts_header!r}"
        ) from exc

    clock = now if now is not None else time.time
    if max_skew_seconds is not None and max_skew_seconds >= 0:
        current = int(clock())
        if abs(current - ts_int) > max_skew_seconds:
            raise TimestampSkew(
                f"{HEADER_TIMESTAMP}: {ts_int} outside ±{max_skew_seconds}s "
                f"of server time {current}"
            )

    signing_input = build_signing_input(method, url, ts_header, body)
    try:
        VerifyKey(pubkey).verify(signing_input, sig)
    except BadSignatureError as exc:
        raise BadSignature("PagerOS-Sig did not verify") from exc

    return VerifiedRequest(device_id=pubkey, timestamp=ts_int)


# --------------------------------------------------------------------------- #
# Signing helper (clients, tests)
# --------------------------------------------------------------------------- #


def sign_request(
    method: str,
    url: str,
    body: bytes | None,
    signing_key: SigningKey,
    *,
    timestamp: int | None = None,
) -> dict[str, str]:
    """Produce ``PagerOS-*`` headers for the given request.

    Mainly intended for tests and for client code that talks to PagerOS
    app servers (the device firmware constructs its own signatures via
    libsodium). On the wire, the same construction is byte-identical.
    """
    if timestamp is None:
        timestamp = int(time.time())
    signing_input = build_signing_input(method, url, timestamp, body)
    sig = signing_key.sign(signing_input).signature
    pubkey = bytes(signing_key.verify_key)
    return {
        HEADER_DEVICE: _b64url_encode(pubkey),
        HEADER_SIG: _b64url_encode(sig),
        HEADER_TIMESTAMP: str(timestamp),
    }


# --------------------------------------------------------------------------- #
# WSGI middleware
# --------------------------------------------------------------------------- #


WSGIEnviron = dict[str, Any]
StartResponse = Callable[[str, list[tuple[str, str]]], Any]
WSGIApp = Callable[[WSGIEnviron, StartResponse], Iterable[bytes]]

ENVIRON_DEVICE_ID = "pageros.device_id"
"""WSGI environ key set on a successfully verified request (bytes pubkey)."""

ENVIRON_TIMESTAMP = "pageros.signed_timestamp"
"""WSGI environ key carrying the request's signed Unix timestamp (int)."""


def _wsgi_headers(environ: WSGIEnviron) -> dict[str, str]:
    """Pull HTTP headers out of a WSGI environ as a plain mapping."""
    out: dict[str, str] = {}
    for key, value in environ.items():
        if key.startswith("HTTP_"):
            header = key[5:].replace("_", "-").title()
            out[header] = value
    return out


def _wsgi_url(environ: WSGIEnviron) -> str:
    """Reconstruct the path-with-query a client signs over."""
    raw_uri = environ.get("RAW_URI") or environ.get("REQUEST_URI")
    if raw_uri:
        return raw_uri
    path = environ.get("PATH_INFO", "")
    query = environ.get("QUERY_STRING", "")
    return f"{path}?{query}" if query else path


def _read_body(environ: WSGIEnviron) -> bytes:
    """Read and re-buffer the request body so downstream apps can re-read it."""
    stream = environ.get("wsgi.input")
    if stream is None:
        return b""
    try:
        length = int(environ.get("CONTENT_LENGTH") or 0)
    except ValueError:
        length = 0
    body = stream.read(length) if length > 0 else b""
    # Put the bytes back so the wrapped app can still read them.
    import io

    environ["wsgi.input"] = io.BytesIO(body)
    return body


class SignatureVerifierMiddleware:
    """WSGI middleware that enforces PagerOS-Sig on every request.

    On success, the wrapped app sees the verified device pubkey in
    ``environ["pageros.device_id"]`` (bytes) and the signed timestamp in
    ``environ["pageros.signed_timestamp"]`` (int).

    On any failure (missing header, malformed encoding, bad signature, or
    timestamp skew), the middleware short-circuits with ``401
    Unauthorized`` and a plain-text body identifying the error class. The
    wrapped app is never called.
    """

    def __init__(
        self,
        app: WSGIApp,
        *,
        max_skew_seconds: int = DEFAULT_MAX_SKEW_SECONDS,
        now: Callable[[], float] | None = None,
    ) -> None:
        self.app = app
        self.max_skew_seconds = max_skew_seconds
        self.now = now

    def __call__(
        self, environ: WSGIEnviron, start_response: StartResponse
    ) -> Iterable[bytes]:
        method = environ.get("REQUEST_METHOD", "GET")
        url = _wsgi_url(environ)
        headers = _wsgi_headers(environ)
        body = _read_body(environ)

        try:
            verified = verify_request(
                method,
                url,
                headers,
                body,
                max_skew_seconds=self.max_skew_seconds,
                now=self.now,
            )
        except SignatureError as exc:
            return _unauthorized(start_response, exc)

        environ[ENVIRON_DEVICE_ID] = verified.device_id
        environ[ENVIRON_TIMESTAMP] = verified.timestamp
        return self.app(environ, start_response)


def _unauthorized(
    start_response: StartResponse, error: SignatureError
) -> list[bytes]:
    reason = type(error).__name__
    body = f"{reason}: {error}".encode("utf-8")
    start_response(
        "401 Unauthorized",
        [
            ("Content-Type", "text/plain; charset=utf-8"),
            ("Content-Length", str(len(body))),
            ("WWW-Authenticate", 'PagerOS-Sig realm="pageros"'),
        ],
    )
    return [body]


__all__ = [
    "BadSignature",
    "DEFAULT_MAX_SKEW_SECONDS",
    "ED25519_PUBKEY_LEN",
    "ED25519_SIG_LEN",
    "ENVIRON_DEVICE_ID",
    "ENVIRON_TIMESTAMP",
    "HEADER_DEVICE",
    "HEADER_SIG",
    "HEADER_TIMESTAMP",
    "InvalidEncoding",
    "MissingHeader",
    "SignatureError",
    "SignatureVerifierMiddleware",
    "TimestampSkew",
    "VerifiedRequest",
    "build_signing_input",
    "compute_body_hash",
    "sign_request",
    "verify_request",
]

File: sdk/js/src/signing.ts
// Request signature verification + signing (JS-004).
//
// Implements the `PagerOS-Sig` request authentication scheme defined in
// SPEC.md §9.2:
//
//     Every request to an App Server includes ``PagerOS-Sig``: an Ed25519
//     signature over ``utf8(method) || utf8(url) || utf8(timestamp) || sha256(body)``.
//     Timestamp prevents replay (servers reject > 5 min skew).
//
// Headers (case-insensitive on input, canonical-cased on output):
//
// - ``PagerOS-Device``    — base64url (no padding) of the device Ed25519
//                           public key (32 bytes).
// - ``PagerOS-Sig``       — base64url (no padding) of the Ed25519 signature
//                           (64 bytes).
// - ``PagerOS-Timestamp`` — integer Unix seconds, as ASCII.
//
// Cryptography uses Node's built-in `node:crypto` Ed25519 support — no
// external deps. Ed25519 signatures are byte-identical to PyNaCl /
// libsodium output, which is what the firmware emits, so wire-level
// vectors round-trip across implementations.

import * as crypto from "node:crypto";

export const HEADER_DEVICE = "PagerOS-Device";
export const HEADER_SIG = "PagerOS-Sig";
export const HEADER_TIMESTAMP = "PagerOS-Timestamp";

export const DEFAULT_MAX_SKEW_SECONDS = 300; // SPEC §9.2

export const ED25519_PUBKEY_LEN = 32;
export const ED25519_SIG_LEN = 64;
export const ED25519_SEED_LEN = 32;

// ---------------------------------------------------------------------------
// Errors
// ---------------------------------------------------------------------------

export class SignatureError extends Error {
  override readonly name: string = "SignatureError";
}
export class MissingHeader extends SignatureError {
  override readonly name = "MissingHeader";
}
export class InvalidEncoding extends SignatureError {
  override readonly name = "InvalidEncoding";
}
export class TimestampSkew extends SignatureError {
  override readonly name = "TimestampSkew";
}
export class BadSignature extends SignatureError {
  override readonly name = "BadSignature";
}

// ---------------------------------------------------------------------------
// Low-level helpers
// ---------------------------------------------------------------------------

const SPKI_ED25519_PREFIX = Buffer.from("302a300506032b6570032100", "hex");
const PKCS8_ED25519_PREFIX = Buffer.from(
  "302e020100300506032b657004220420",
  "hex",
);

export interface VerifiedRequest {
  /** Raw 32-byte Ed25519 device public key (the PagerOS device identity). */
  readonly deviceId: Uint8Array;
  /** The Unix-seconds timestamp embedded in the request. */
  readonly timestamp: number;
}

export function computeBodyHash(body: Uint8Array | null | undefined): Uint8Array {
  const hash = crypto.createHash("sha256");
  if (body && body.length > 0) hash.update(body);
  return new Uint8Array(hash.digest());
}

export function buildSigningInput(
  method: string,
  url: string,
  timestamp: string | number,
  body: Uint8Array | null | undefined,
): Uint8Array {
  const ts = String(timestamp);
  const enc = new TextEncoder();
  const m = enc.encode(method);
  const u = enc.encode(url);
  const t = enc.encode(ts);
  const h = computeBodyHash(body);
  const out = new Uint8Array(m.length + u.length + t.length + h.length);
  out.set(m, 0);
  out.set(u, m.length);
  out.set(t, m.length + u.length);
  out.set(h, m.length + u.length + t.length);
  return out;
}

function base64UrlDecode(value: string, field: string, expectedLen: number): Uint8Array {
  if (typeof value !== "string") {
    throw new InvalidEncoding(`${field}: not a string`);
  }
  let buf: Buffer;
  try {
    buf = Buffer.from(value, "base64url");
  } catch (err) {
    throw new InvalidEncoding(`${field}: not valid base64url (${(err as Error).message})`);
  }
  if (buf.length !== expectedLen) {
    throw new InvalidEncoding(
      `${field}: expected ${expectedLen} bytes, got ${buf.length}`,
    );
  }
  return new Uint8Array(buf.buffer, buf.byteOffset, buf.byteLength);
}

function base64UrlEncode(raw: Uint8Array): string {
  return Buffer.from(raw).toString("base64url");
}

function getCaseInsensitive(
  headers: Record<string, string>,
  key: string,
): string | undefined {
  if (key in headers) return headers[key];
  const lower = key.toLowerCase();
  for (const k of Object.keys(headers)) {
    if (k.toLowerCase() === lower) return headers[k];
  }
  return undefined;
}

function ed25519PublicKey(raw: Uint8Array): crypto.KeyObject {
  if (raw.length !== ED25519_PUBKEY_LEN) {
    throw new InvalidEncoding(
      `ed25519 public key: expected ${ED25519_PUBKEY_LEN} bytes, got ${raw.length}`,
    );
  }
  const der = Buffer.concat([SPKI_ED25519_PREFIX, Buffer.from(raw)]);
  return crypto.createPublicKey({ key: der, format: "der", type: "spki" });
}

function ed25519PrivateKey(seed: Uint8Array): crypto.KeyObject {
  if (seed.length !== ED25519_SEED_LEN) {
    throw new InvalidEncoding(
      `ed25519 seed: expected ${ED25519_SEED_LEN} bytes, got ${seed.length}`,
    );
  }
  const der = Buffer.concat([PKCS8_ED25519_PREFIX, Buffer.from(seed)]);
  return crypto.createPrivateKey({ key: der, format: "der", type: "pkcs8" });
}

/** Derive the 32-byte Ed25519 public key for a 32-byte seed. */
export function publicKeyFromSeed(seed: Uint8Array): Uint8Array {
  const priv = ed25519PrivateKey(seed);
  const pub = crypto.createPublicKey(priv);
  // Extract raw 32-byte key by exporting as JWK.
  const jwk = pub.export({ format: "jwk" }) as { x?: string };
  if (typeof jwk.x !== "string") {
    throw new Error("ed25519 public key export missing 'x'");
  }
  return new Uint8Array(Buffer.from(jwk.x, "base64url"));
}

// ---------------------------------------------------------------------------
// Verification
// ---------------------------------------------------------------------------

export interface VerifyRequestOptions {
  method: string;
  url: string;
  headers: Record<string, string>;
  body?: Uint8Array | null;
  /**
   * Reject requests whose ``PagerOS-Timestamp`` is more than this many
   * seconds away from ``now()``. Set to ``null`` to disable the skew
   * check (e.g. in offline LoRa-path tests where the device clock is not
   * authoritative).
   */
  maxSkewSeconds?: number | null;
  now?: () => number; // returns Unix seconds (float OK)
}

export function verifyRequest(opts: VerifyRequestOptions): VerifiedRequest {
  const { method, url, headers, body = null } = opts;
  const maxSkew = opts.maxSkewSeconds === undefined
    ? DEFAULT_MAX_SKEW_SECONDS
    : opts.maxSkewSeconds;
  const now = opts.now ?? (() => Date.now() / 1000);

  const deviceHeader = getCaseInsensitive(headers, HEADER_DEVICE);
  const sigHeader = getCaseInsensitive(headers, HEADER_SIG);
  const tsHeader = getCaseInsensitive(headers, HEADER_TIMESTAMP);

  if (deviceHeader === undefined) throw new MissingHeader(`missing ${HEADER_DEVICE}`);
  if (sigHeader === undefined) throw new MissingHeader(`missing ${HEADER_SIG}`);
  if (tsHeader === undefined) throw new MissingHeader(`missing ${HEADER_TIMESTAMP}`);

  const pubkey = base64UrlDecode(deviceHeader, HEADER_DEVICE, ED25519_PUBKEY_LEN);
  const sig = base64UrlDecode(sigHeader, HEADER_SIG, ED25519_SIG_LEN);

  const tsInt = Number(tsHeader);
  if (!Number.isFinite(tsInt) || !Number.isInteger(tsInt) || !/^-?\d+$/.test(tsHeader)) {
    throw new InvalidEncoding(
      `${HEADER_TIMESTAMP}: expected integer Unix seconds, got ${JSON.stringify(tsHeader)}`,
    );
  }

  if (maxSkew !== null && maxSkew >= 0) {
    const current = Math.floor(now());
    if (Math.abs(current - tsInt) > maxSkew) {
      throw new TimestampSkew(
        `${HEADER_TIMESTAMP}: ${tsInt} outside ±${maxSkew}s of server time ${current}`,
      );
    }
  }

  const signingInput = buildSigningInput(method, url, tsHeader, body);
  const keyObj = ed25519PublicKey(pubkey);
  const ok = crypto.verify(null, signingInput, keyObj, sig);
  if (!ok) {
    throw new BadSignature("PagerOS-Sig did not verify");
  }
  return { deviceId: pubkey, timestamp: tsInt };
}

// ---------------------------------------------------------------------------
// Signing helper (clients, tests)
// ---------------------------------------------------------------------------

export interface SignRequestOptions {
  method: string;
  url: string;
  body?: Uint8Array | null;
  /** 32-byte Ed25519 seed (the same input PyNaCl / libsodium accept). */
  seed: Uint8Array;
  timestamp?: number;
}

export function signRequest(opts: SignRequestOptions): Record<string, string> {
  const { method, url, body = null, seed } = opts;
  const timestamp = opts.timestamp ?? Math.floor(Date.now() / 1000);

  const priv = ed25519PrivateKey(seed);
  const signingInput = buildSigningInput(method, url, timestamp, body);
  const sig = crypto.sign(null, signingInput, priv);
  const pub = publicKeyFromSeed(seed);
  return {
    [HEADER_DEVICE]: base64UrlEncode(pub),
    [HEADER_SIG]: base64UrlEncode(new Uint8Array(sig)),
    [HEADER_TIMESTAMP]: String(timestamp),
  };
}
